include/lbann/execution_algorithms/ltfb/random_pairwise_exchange.hpp

```diff
@@ -0,0 +1,268 @@
+   *                 output, the winning model with respect to the
+   *                 tournament.
+   *  @param[in,out] ctxt The execution context for the outer LTFB
+   *                 wrapper.
+   *  @param[in,out] dc The data source for the tournament.
+   */
+  void select_next(model& m,
+                   ltfb::ExecutionContext& ctxt,
+                   data_coordinator& dc) const final;
+
+private:
+  /** @brief Get the value of the given metric from the model. */
+  EvalType
+  evaluate_model(model& m, ExecutionContext& ctxt, data_coordinator& dc) const;
+  /** @brief Generate a new trainer partner from the comm. */
+  El::Int get_partner_trainer(lbann_comm const& c) const noexcept;
+  /** @brief Evaluate the output of two models according to the input
+   *         metric strategy.
+   *
+   *  The local model is preferred if possible (to avoid a model
+   *  move). That is, "<=" or ">=" is used.
+   *
+   *  @param[in] local The metric output of the local model.
+   *  @param[in] remote The metric output of the remote model.
+   *  @returns true if the local model is favored.
+   */
+  bool local_is_better(EvalType local, EvalType remote) const;
+
+private:
+  /** @brief Strategy to decide if the local or remote value is better. */
+  metric_strategy m_metric_strategy;
+  /** @brief The name of the metric to be used to evaluate a model.
+   *
+   *  A metric with the given name must exist in every model seen.
+   */
+  std::string m_metric_name;
+  /** @brief The strategy for exchanging two models.
+   *
+   *  This is largely an implementation detail of moving models
+   *  around. It shouldn't be used to alter the algorithmic goings-on
+   *  of the method. Ideally, it should disappear as we converge on a
+   *  "universally best method" for moving things around (let's
+   *  pretend such a thing exists; it makes me feel better, anyway).
+   */
+  std::unique_ptr<ExchangeStrategy> m_comm_algo;
+
+}; // class RandomPairwiseExchange
+
+/** @class SendRecvWeights
+ *  @brief Exchange model weights directly using sendrecvs.
+ *  @todo More general approach to exchange optimizer state. Currently
+ *  only SGD and Adam are supported.
+ */
+class SendRecvWeights final
+  : public Cloneable<SendRecvWeights, RandomPairwiseExchange::ExchangeStrategy>
+{
+  using BaseType =
+    Cloneable<SendRecvWeights, RandomPairwiseExchange::ExchangeStrategy>;
+
+public:
+  /** @brief Construct from weights names
+   *  @param[in] weights_names Names of weights to exchange. If empty,
+   *                           then all weights are exchanged.
+   *  @param[in] exchange_hyperparameters Exchange optimizer
+   *                                      hyperparameters.
+   */
+  SendRecvWeights(std::set<std::string> const& weights_names,
+                  bool exchange_hyperparameters);
+
+  /** @brief Construct from weights names
+   *  @param[in] weights_names Names of weights to exchange. If empty,
```

Comment:
Part of unit test (if any) should cover this (that is sendrcv of selected weights)